### PR TITLE
User-Agent headers added to each request

### PR DESCRIPTION
**Pull request by beevital**
Related to [T-135809](https://fivetran.height.app/T-135809)

To provide information about our API usage we should add User-Agent headers to requests to identify requests from Go-Fivetran and Terraform.

Client.UserAgent method allows to set-up custom User-Agent header - this method will be used by terraform provider.
By default header will be: "User-Agent":"Go-Fivetran/<version>" 

changes tested on local environment via examples.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -9,20 +9,37 @@ import (
...
+// UserAgent sets custom User-Agent header in Client requests
+func (c *Client) UserAgent(userAgent string) {
+	c.userAgent = userAgent
+}
+
+func (c *Client) fillHeaders() map[string]string {
+	headers := make(map[string]string)
+	headers["Authorization"] = c.authorization
+	headers["User-Agent"] = c.userAgent
+	return headers
```
Comment (shirolimit) [inline client.go:44]: Just curious if we could use:

```go
return map[string]string{
    "Authorization": c.authorization,
    "User-Agent": c.userAgent,
}
```

File: client.go [hunk 2]
```diff
@@ -9,20 +9,37 @@ import (
...
 	baseURL       string
 	authorization string
+	userAgent     string
 }
 
 const defaultBaseURL = "https://api.fivetran.com/v1"
 const restAPIv2 = "application/json;version=2"
 
+// WARNING: Update Agent version on each release!
+const defaultUserAgent = "Go-Fivetran/0.2.3"
```
Comment (shirolimit) [inline client.go:19]: We might want to have a separate constant with version only.
But it is no necessary at the moment.

**Review by shirolimit — DISMISSED** [review]
LGTM! 👍

File: client.go (hunk 2 above)
Comment (beevital) [reply client.go:19]: I think it's not a problem to update version here=) It's the only place we use version in code

**Review by felipeneuwald — CHANGES_REQUESTED** [review]
- Wouldn't be easier to just change `func httpRequest`? 
https://github.com/fivetran/go-fivetran/blob/main/http_request.go#L19

- Missing CHANGELOG

**Comment by beevital** [comment]
> * Wouldn't be easier just changing `func httpRequest`?
>   https://github.com/fivetran/go-fivetran/blob/main/http_request.go#L19
> * Missing CHANGELOG

We should pass user-agent value from client somehow to http_request. I think http_request function dont know anything about client, or any other specifics - it's just make http requests.

Will add the changelog

**Comment by felipeneuwald** [comment]
> > * Wouldn't be easier just changing `func httpRequest`?
> >   https://github.com/fivetran/go-fivetran/blob/main/http_request.go#L19
> > * Missing CHANGELOG
> 
> We should pass user-agent value from client somehow to http_request. I think http_request function doesn't know anything about client, or any other specifics - it's just make http requests.
> 
> Will add the changelog

Also, how `terraform-provider-fivetran` will pass user-agent data to `go-fivetran`? IMO it would be good to have a way to receive custom user-agent data and append that custom value to `defaultUserAgent` (or something like `defaultUserAgent + customUserAgent`).

**Comment by beevital** [comment]
> > > * Wouldn't be easier just changing `func httpRequest`?
> > >   https://github.com/fivetran/go-fivetran/blob/main/http_request.go#L19
> > > * Missing CHANGELOG
> > 
> > 
> > We should pass user-agent value from client somehow to http_request. I think http_request function doesn't know anything about client, or any other specifics - it's just make http requests.
> > Will add the changelog
> 
> Also, how `terraform-provider-fivetran` will pass user-agent data to `go-fivetran`? IMO it would be good to have a way to receive custom user-agent data and append that custom value to `defaultUserAgent` (or something like `defaultUserAgent + customUserAgent`).

For this purpose - UserAgent method defined at client. When creating client in terraform-provider - you can call client.UserAgent(Go-Fivetran/<version> terraform-provider) for example.

**Comment by beevital** [comment]
> > > > * Wouldn't be easier just changing `func httpRequest`?
> > > >   https://github.com/fivetran/go-fivetran/blob/main/http_request.go#L19
> > > > * Missing CHANGELOG
> > > 
> > > 
> > > We should pass user-agent value from client somehow to http_request. I think http_request function doesn't know anything about client, or any other specifics - it's just make http requests.
> > > Will add the changelog
> > 
> > 
> > Also, how `terraform-provider-fivetran` will pass user-agent data to `go-fivetran`? IMO it would be good to have a way to receive custom user-agent data and append that custom value to `defaultUserAgent` (or something like `defaultUserAgent + customUserAgent`).
> 
> For this purpose - UserAgent method defined at client. When creating client in terraform-provider - you can call client.UserAgent(Go-Fivetran/ terraform-provider) for example.

You can just override the userAgent field value with UserAgent method. I haven't modified TF-provider (we can make it only after release of go-fivetran)

**Comment by felipeneuwald** [comment]
> > > > > * Wouldn't be easier just changing `func httpRequest`?
> > > > >   https://github.com/fivetran/go-fivetran/blob/main/http_request.go#L19
> > > > > * Missing CHANGELOG
> > > > 
> > > > 
> > > > We should pass user-agent value from client somehow to http_request. I think http_request function doesn't know anything about client, or any other specifics - it's just make http requests.
> > > > Will add the changelog
> > > 
> > > 
> > > Also, how `terraform-provider-fivetran` will pass user-agent data to `go-fivetran`? IMO it would be good to have a way to receive custom user-agent data and append that custom value to `defaultUserAgent` (or something like `defaultUserAgent + customUserAgent`).
> > 
> > 
> > For this purpose - UserAgent method defined at client. When creating client in terraform-provider - you can call client.UserAgent(Go-Fivetran/ terraform-provider) for example.
> 
> You can just override the userAgent field value with UserAgent method. I haven't modified TF-provider (we can make it only after release of go-fivetran)

I see. So, what about changing `"User-Agent":    c.userAgent,`:

```go
func (c *Client) commonHeaders() map[string]string {
	return map[string]string{
		"Authorization": c.authorization,
		"User-Agent":    c.userAgent,
	}
}
```
to something like `"User-Agent":    defaultUserAgent + ";" + c.userAgent,`? Of course with some extra logic to only add a `;` if there's a `c.UserAgent`. WDYT?

Also, I suggest changing `c.userAgent` to `c.customUserAgent`; so then we have `defaultUserAgent` and `customUserAgent`.

File: CHANGELOG.md [hunk 3]
```diff
@@ -5,7 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-## [Unreleased](https://github.com/fivetran/go-fivetran/compare/v0.2.1...HEAD)
+## [Unreleased](https://github.com/fivetran/go-fivetran/compare/v0.2.3...HEAD)
+
+## [0.2.3](https://github.com/fivetran/go-fivetran/compare/v0.2.2...v0.2.3) - 2021-11-10
```
Comment (felipeneuwald) [inline CHANGELOG.md:10]: I think this is a new feature and not a fix. What about version `0.3.0`?

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,7 +5,12 @@ All notable changes to this project will be documented in this file.
...
+
+### Added
+- `CustomUserAgent` method for overriding User-Agent header in http-responses (for terraform-provider)
```
Comment (felipeneuwald) [inline CHANGELOG.md:13]: Not only for `terraform-provider-fivetran`, but for any other application using the SDK.

File: client.go [hunk 5]
```diff
@@ -7,22 +7,45 @@ import (
...
+		customUserAgent: "",
 	}
 }
 
 // BaseURL changes Client base REST API endpoint URL
 func (c *Client) BaseURL(baseURL string) {
 	c.baseURL = baseURL
 }
+
+// UserAgent sets custom User-Agent header in Client requests
```
Comment (felipeneuwald) [inline client.go:35]: Replace // UserAgent by // CustomUserAgent

File: client.go [hunk 6]
```diff
@@ -7,22 +7,45 @@ import (
...
+const defaultUserAgent = "Go-Fivetran/0.2.3"
+
 // New receives API Key and API Secret, and returns a new Client
 func New(apiKey, apiSecret string) *Client {
 	return &Client{
-		baseURL:       defaultBaseURL,
-		authorization: fmt.Sprintf("Basic %v", base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%v:%v", apiKey, apiSecret)))),
+		baseURL:         defaultBaseURL,
+		authorization:   fmt.Sprintf("Basic %v", base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%v:%v", apiKey, apiSecret)))),
+		customUserAgent: "",
```
Comment (felipeneuwald) [inline client.go:26]: Not needed. Default string type zero value is an empty string.

File: client.go [hunk 7]
```diff
@@ -7,22 +7,45 @@ import (
...
+	c.customUserAgent = customUserAgent
+}
+
+func (c *Client) commonHeaders() map[string]string {
+	userAgent := defaultUserAgent
+
+	if c.customUserAgent != "" {
+		userAgent += " " + c.customUserAgent
+	}
+
```
Comment (felipeneuwald) [inline client.go:46]: I think the map needs to be initialized first. 
```go
commonHeaders := make(map[string]string)
...
return commonHeaders
}
```
```

**Review by felipeneuwald — CHANGES_REQUESTED** [review]
Some changes were proposed.

File: CHANGELOG.md (hunk 3 above)
Comment (beevital) [reply CHANGELOG.md:10]: Ok, let it be so)

File: client.go (hunk 7 above)
Comment (beevital) [reply client.go:46]: @shirolimit proposed the opposit approach, initially I've done this with make

**Review by felipeneuwald — APPROVED** [review]
LGTM

**Pull request merged**